### PR TITLE
TC-IDM-10.2: Add check for MACL

### PR DESCRIPTION
**Pull request by cecille**
Test: unit tests. Note this has not been tested against the example
      app as it is still under development.

### CODE REVIEW

**Review by ksperling-apple — APPROVED** [review]

File: src/python_testing/TC_DeviceConformance.py [hunk 1]
```diff
@@ -50,6 +50,22 @@ async def setup_class_helper(self):
...
         self.problems.extend(problems)
 
+    def _get_device_type_id(self, device_type_name: str) -> int:
+        id = [id for id, dt in self.xml_device_types.items() if dt.name.lower() == device_type_name.lower()]
+        if len(id) != 1:
+            self.fail_current_test(f"Unable to find {device_type_name} device type")
+        return id[0]
+
+    def _has_nim(self):
+        nim_id = self._get_device_type_id('network infrastructure manager')
```
Comment (tcarmelveilleux) [inline src/python_testing/TC_DeviceConformance.py:60]: For `_get_device_type_id` handle case normalization? Would the call `_get_device_type_id('Network Infrastructure Manager')` also work?

File: src/python_testing/TestConformanceTest.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        root_endpoint[ac.id][ac.Attributes.FeatureMap.attribute_id] = ac.Bitmaps.Feature.kManagedDevice
+        root_endpoint[ac.id][ac.Attributes.Arl.attribute_id] = []
+        root_endpoint[ac.id][ac.Attributes.CommissioningARL.attribute_id] = []
+        root_endpoint[ac.id][ac.Attributes.AttributeList.attribute_id].extend([
+            ac.Attributes.Arl.attribute_id, ac.Attributes.CommissioningARL.attribute_id])
+        root_endpoint[ac.id][ac.Attributes.AcceptedCommandList.attribute_id].append(ac.Commands.ReviewFabricRestrictions.command_id)
+        root_endpoint[ac.id][ac.Attributes.GeneratedCommandList.attribute_id].append(
+            ac.Commands.ReviewFabricRestrictionsResponse.command_id)
+
+    @ async_test_body
```
Comment (tcarmelveilleux) [inline src/python_testing/TestConformanceTest.py:100]: It's odd that there is a space here between the `@` and `async_test_body`

File: src/python_testing/TC_DeviceConformance.py [hunk 3]
```diff
@@ -50,6 +50,22 @@ async def setup_class_helper(self):
...
+        for endpoint in self.endpoints_tlv.values():
+            desc = Clusters.Descriptor
+            device_types = [dt.deviceType for dt in endpoint[desc.id][desc.Attributes.DeviceTypeList.attribute_id]]
+            if nim_id in device_types:
+                # TODO: it's unclear if this needs to be present on every endpoint. Right now, this assumes one is sufficient.
```
Comment (tcarmelveilleux) [inline src/python_testing/TC_DeviceConformance.py:65]: Presence of only 1 is OK, since it's an application device type.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/python_testing/TC_DeviceConformance.py (hunk 1 above)
Comment (cecille) [reply src/python_testing/TC_DeviceConformance.py:60]: yup

File: src/python_testing/TC_DeviceConformance.py (hunk 3 above)
Comment (cecille) [reply src/python_testing/TC_DeviceConformance.py:65]: The question was around multiple endpoints. So one endpoint with NIM one with ex. lightbulb. I'm not sure how the spec text is going to land around that condition.

File: src/python_testing/TestConformanceTest.py (hunk 2 above)
Comment (cecille) [inline src/python_testing/TestConformanceTest.py:100]:
```suggestion
    @async_test_body
```

Comment (cecille) [reply src/python_testing/TestConformanceTest.py]: yeah, that is odd. Fixed.

File: src/python_testing/TC_DeviceConformance.py (hunk 3 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceConformance.py:65]: Given ACL is node scoped and the fact that we do not want to allow MACL for other device types than NIM, I would say that we have a requirement that NIM is not allowed to have other EP's with other device types 🤔

Comment (tcarmelveilleux) [reply src/python_testing/TC_DeviceConformance.py:65]: Let's not throw the baby out with the bathwater. It makes no sense saying there can only be a single endpoint and it has to be NIM.

Maybe we need to update ACL TP instead and work an update on the feature conformance to add steps that it is only present if at least one endpoint is a device type that has it as O at least.

Comment (cecille) [reply src/python_testing/TC_DeviceConformance.py:65]: This does this currently - allowed if at least one endpoint is marked as being a NIM. It does mean you can have a weird combo device, but you're probably right that it's unlikely to happen in practice.

File: src/python_testing/TestConformanceTest.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        self.problems.extend(problems)
+        asserts.assert_true(success, "Unexpected failure with NIM and MACL")
+
+        # A MACL is not allowed when there is no NIM
+        self.endpoints_tlv[1] = self._create_minimal_dt(device_type_id=on_off_id)
+        success, problems = self.check_conformance(ignore_in_progress=False, is_ci=False)
+        self.problems.extend(problems)
+        asserts.assert_false(success, "Unexpected success with On/Off and MACL")
+
+        # TODO: what happens if there is a NIM and a non-NIM endpoint?
```
Comment (chrisdecenzo) [inline src/python_testing/TestConformanceTest.py:127]: That is allowed. The ARL entries would only be allowed on the clusters of the NIM endpoint

File: src/python_testing/TC_DeviceConformance.py [hunk 5]
```diff
@@ -125,6 +141,13 @@ def record_warning(location, problem):
                 for f in feature_masks:
                     location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id,
                                                      attribute_id=GlobalAttributeIds.FEATURE_MAP_ID)
+                    if cluster_id == Clusters.AccessControl.id and f == Clusters.AccessControl.Bitmaps.Feature.kManagedDevice:
+                        # Managed ACL is treated as a special case because it is only allowed if other endpoints support NIM and disallowed otherwise.
+                        if not self._has_nim():
```
Comment (chrisdecenzo) [inline src/python_testing/TC_DeviceConformance.py:146]: Given that we will likely add to this list in the future, shouldn't this be self._has_allowed_MACL()?

Do we have hard coding of the device library rules in the other scripts?

File: src/python_testing/TC_DeviceConformance.py (hunk 3 above)
Comment (cecille) [reply src/python_testing/TC_DeviceConformance.py:65]: OK, I'm going to resolve this for now so I can get this code in. I've opened a spec issue here: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/10294

**Review by chrisdecenzo — CHANGES_REQUESTED** [review]
Can we do this in a more generic way then hardcoding NIM? Seems like we should be checking against a list of device types that are allowed to use this feature.

File: src/python_testing/TestConformanceTest.py (hunk 4 above)
Comment (ReneJosefsen) [reply src/python_testing/TestConformanceTest.py:127]: Where in the spec is this stated? The ACL cluster is node scopes so it applies to all the EP's on the device and I have not found any such restrictions mentioned in the spec?

**Pull request merged**